docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+This gives us parity with the current system while allowing us to iteratively assign more accurate and descriptive types, giving us the ability to catch more errors at compile time.
+
+Below I attempt to assign accurate types to common Flux transformations.
+
+* from
+
+    ```
+    forall ['r, 's] where 's: StrRec (org: str, bucket: str) -> [ {_time: time | _start: time | _stop: time | _value: 'r | 's} ]
+    ```
+    This polymorphic type signature says that `from` is a function that returns a list of records.
+    These records have fields `_time`, `_start`, `_stop`, and `_value`.
+    The `_value` field can be of any type.
+    All other fields represented by the row variable `'s` must be of type _string_.
+    Note the last part corresponds to the tag values of a series.
+
+    An example of a program that will fail to compile is:
+    ```
+    from(bucket: "telegraf")
+        |> range(start: -1h)
+        |> filter(fn: (r) => r.host == 1) // error: {host: int | 't} is not of kind StrRec
+    ```
+    The reason for failed compilation will be clear soon, but basically `range` and `filter` are just pass through transformations.
+    They don't extend or modify the types of their inputs.
+    As a result `filter` expects to receive the same list type that `from` returns and so it expects `r.host` to be of type _string_.
+    This contradicts the usage of `r.host` as an _int_.
+
+* range
+
+    ```
+    forall ['r] where 'r:Rec (start: time, stop: time, tables: [ {_start: time | _stop: time | 'r} ]) -> [ {_start: time | _stop: time | 'r} ]
+    ```
+    Let's focus on the `tables` parameter.
+    `range` takes a `tables` parameter that is a list of records that must have fields `_start` and `_stop` and returns the same list type.
+
+* filter
+
+    ```
+    forall ['r] where 'r:Rec (fn: (r: 'r) -> bool, tables: ['r]) -> ['r]
+    ```
+    `filter` takes a list of records and returns that same list of records.
+
+    From the type signatures of `range` and `filter` it should be clear why the above example program fails unification.
+
+* map
+
+    ```
+    forall ['r, 's] where 'r:Rec, 's:Rec (fn: ('r) => 's, tables: [ 'r ]) -> [ 's ]
+    ```
+
+* keep
+
+    `keep` and `drop` operations can be implemented using `map`:
+    ```
+    map(fn: (r) => ( {a: r.a | b: r.b | {} } )) // keep columns a and b
+    map(fn: (r) => ( {r - a, b} ))              // drop columns a and b
+    ```
+    Assigning a type signature to the actual `keep` and `drop` functions unfortunatly is not so simple.
+    These functions take the names of the columns they operate on via a `columns` parameter.
+    Assigning an accurate type signature to these functions requires the notion of compile time constants or label types [#1122](https://github.com/influxdata/flux/issues/1122).
+
+    Another option is a preprocessing step before type inference that rewrites the Flux AST by replacing calls to `keep` and `drop` with their `map` equivalents.
```

Comment:
I see `label` types as a sort of preprocessing step anyways, so this fits. 

Tangential but I think we should actually implement `keep` and `drop` in pure Flux and add a planner rule to rewrite them to their specific transformations. 

Using psuedo syntax  from label types here are the `keep` and `drop` type signatures, these signatures assume only a single column is applied at once. In the case of multiple columns it can be seen as multiple calls to the function.

```
keep : forall ['r] where 'r:Rec  (column: label, tables: [ 'r ]) -> [ {[label] : r[label] } ]
drop : forall ['r] where 'r:Rec  (column: label, tables: [ 'r ]) -> [ { r - [label] } ]
```